unix like redirect output

meeuw opened the pull request:
## Description
Hi, could you please give me some feedback on this PR? (It might be a fix for issue #91)
I need to write some unit tests to improve the code coverage.

Output from a query can be redirected like:
mysql root@172.17.0.2:(none)> select 1 > /tmp/output

Output from a query can be appended like:
mysql root@172.17.0.2:(none)> select 2 >> /tmp/output

You can also use spaces for redirection like:
mysql root@172.17.0.2:(none)> select 1 > "/tmp/output 123"

And even escape quotes:
mysql root@172.17.0.2:(none)> select 1 > "/tmp/\\"weird\\""

After a query the redirection is automatically disabled. The complete output is written to a file, including the "1 row in set"

## Checklist
- [x] I've added this contribution to the `changelog.md`.
- [x] I've added my name to the `AUTHORS` file (or it's already there).

tsroten commented:
I really like this idea! I would definitely use this. I'll post some feedback on some of the code.

tsroten commented on mycli/packages/special/iocommands.py:
```diff
@@ -240,3 +240,54 @@ def execute_system_command(arg, **_):
+        if m:
+            filename = m.group(1)
+        else:
+            filename = None
```
Both here and on line 268, you set `filename` to `None`. You can set it to `None` before the conditionals so you don't have to set it twice.

meeuw replied:
fixed :-)

tsroten commented on mycli/packages/special/iocommands.py:
```diff
@@ -240,3 +240,50 @@ def execute_system_command(arg, **_):
+    filename = None
+    if sql[-1] in ("'", '"'): # find filename in quoted redirect
+        quote = sql[-1]
+        m = re.match(r'.*([^\\]'+quote+r'.*?'+quote+'$)', sql)
```
Rather than putting the parenthesis around the quotes and the filename, put them just around the filename (inside the quotes), so that you don't have to remove the quotes later on. Then, `filename = m.group(1)` won't include the quotes.

Then you can get rid of line 262 (`quote = ''`) and lines 277-278 (`if quote: filename = filename[len(quote):-len(quote)]`).

Side note: if possible, put spaces around operators like `+`, e.g. `re.match(foo + bar)` instead of `re.match(foo+bar)`. It makes it easier to read.

tsroten replied:
Hmm, actually I say that, but then I see that you use the length of the filename in line 267 to remove it from the query. That wouldn't work with my suggestion.

tsroten commented on mycli/main.py:
```diff
@@ -730,6 +733,15 @@ def run_query(self, query, table_format=None, new_line=True):
                 click.echo(line, nl=new_line)
 
+    def output_to_file(self, text):
+        output = special.get_redirected_output()
```
This is the second time `get_redirected_output()` gets called (see line 596). Can we pass an `append` argument to this function so we don't have to call it again?

tsroten replied:
Gosh, I need to drink my coffee for the day! I didn't notice it was storing the response (`use_redirected_output`) instead of computing it again. Feel free to ignore that. Only suggestion here might be to condense the conditional `mode = 'a' if output['append'] else 'w'`.

tsroten commented:
I'll test it later on today, right now I've got to run. It would make the code more readable if you put newlines between conditional blocks.

So, instead of:
```python
if:
  foo
else:
  bar
if:
  oof
else:
  rab
```

Do this:
```python
if:
  foo
else:
  bar

if:
  oof
else:
  rab
```

meeuw commented:
@tsroten many thanks for the feedback

tsroten commented:
After testing this feature, here is some feedback:

1. This feature is very helpful. I've been wanting this for awhile.
2. The bash-like syntax of `>` has serious issues with SQL comparison operators. E.g. try to include a greater-than comparison in a where clause (`select * from foo where id > 100;`).
3. I wonder if a tee-like feature would be more handy than a strict redirect. In other words, I think it might be better to both redirect *and* display the output in the client (see mysql's `tee` command](https://dev.mysql.com/doc/refman/5.7/en/mysql-commands.html)). Otherwise, you can't ensure you are getting the correct data.
4. My guess is that it would be better to implement this using a `\` syntax like mycli's existing commands. For example, something like `\tee [file] [query]` or `\tee [query] [file]`. Or, we need to come up with a better syntax that won't clash with SQL.

I'd be interested in the rest of the team's thoughts, @dbcli/mycli-core.

j-bennet commented:
I agree with @tsroten on #4. I think it's better to not mix bash with sql, so +1 on making it into special command syntax.

`tee` looks like it would be a great addition to `mycli`. In mysql, `tee [filename]` sets file name to dump everything into (queries and output), and `untee` unsets it. We could add a flag to mycli's `\tee` to dump results only (`tee [filename] --results-only`).

darikg commented:
+1 with @j-bennet and @tsroten on using a `\special` command rather than redirect operators

meeuw commented:
Bah I forgot about the greater-than operator of SQL, thanks for spotting this.

I will rewrite this PR to use \>> and \> operators and create a new PR to use a \tee special command (which wouldn't require the hacks in sqlexecute). Then we can check what works better.

meeuw commented:
I've changed it to \>> and \>

meeuw commented:
Allright, I've changed the > and >> operators to \\> and \\>> so I guess it doesn't conflict with SQL operators anymore. In the end this is quite different from the tee implementation because tee adds the sql commands in the output and isn't automatically disabled. 

I'd like to use this to export query output to tsv or json, I suppose I only have to strip the n row(s) in set for maximum usability.

tsroten commented on mycli/main.py:
```diff
@@ -728,6 +731,12 @@ def run_query(self, query, table_format=None, new_line=True):
+    def output_to_file(self, text):
+        output = special.get_redirected_output()
+        mode = 'a' if output['append'] else 'w'
+        with open(output["filename"], mode) as f:
```
We should probably open with `encoding='utf8'`

tsroten commented:
@meeuw Okay, let's get this edited and merged in :) It's been sitting here awhile. Sorry about that!

Here are two thoughts:

1. Can we move the `output_to_file` to the special file so that the actual writing is out of `main.py` like you did in the tee command? Basically, let's make it consistent.

2. I think we can refactor the regular expression matching quite a bit. Here is an idea you might want to try:

```python
import re

sql = "SELECT first_name, last_name FROM customers \>> '../my file.txt'"
m = re.match(r'(?P<sql>.*)(?P<operator>\\>{1,2})\s+(?P<quote>[\'"]?)(?P<filename>.*)(?P=quote)', sql)

if m:
    append = m.group('operator').endswith('>>')
    print(m.group('sql'))
    print(m.group('operator'))
    print("Append: {}".format(append or 'False'))
    print(m.group('filename'))
else:
    print('No redirect match')
```

Outputs:
```
SELECT first_name, last_name FROM customers
Append: True
\>>
../my file.txt
```

meeuw commented:
@tsroten thanks for your feedback! I've moved output_to_file to iocommand.py and added utf-8 encoding.

The current regular expressions are indeed complicated but I couldn't find another way of verifying nested quotes.

For example, I can't find a single regular expression to pass (nested quotation):
```sql = "SELECT first_name, last_name, \" \>> \" FROM customers"```

Another problem is that the quotation cannot be checked (non matching quotation):
```sql = "SELECT first_name, last_name FROM customers \>> '../my file.txt\""```

As an alternative maybe we could use https://docs.python.org/3/library/shlex.html but I don't think that would make it less complicated and I'm not sure if SQL escaped strings are compatible with shell script.

Shell script is very flexible with opening/closing quotes:
```
>>> shlex.split('""""')
['']
```

And mycli/mysql isn't
```
mysql root@172.17.0.2:(none)> select """"
+---+
| " |
+---+
| " |
+---+
1 row in set
Time: 0.002s
```

tsroten commented:
Thanks for the good examples. 👍 It looks like you've thought through some of the edge cases.

I'm interested in your thoughts on whether or not this is forcing SQL into places it shouldn't be (e.g. mixed with BASH-like syntax).

pgcli does it with a [`\o` special command](https://github.com/dbcli/pgcli/pull/483) that toggles output on/off. That would match the way our `\tee` command works.

I am not opposed to adding the redirect command, I'm just curious on your thoughts on how it fits within a database REPL.

amjith commented:
> I'm interested in your thoughts on whether or not this is forcing SQL into places it shouldn't be (e.g. mixed with BASH-like syntax).

I've been having the exact same thought. It seems like a harmless addition that adds functionality but I've been having mixed feelings about adding a `>>` syntax to SQL. In some ways, I feel like it is intuitive and wish that SQL had this feature built-in. But the fact that it doesn't have this built-in makes me wonder if this feature will get enough adoption or if this will be an esoteric feature that only the devs of the tool know about.

meeuw commented:
> I'm interested in your thoughts on whether or not this is forcing SQL into places it shouldn't be (e.g. mixed with BASH-like syntax).

Well, with a special command it is outside the SQL language. The advantage of the BASH-like syntax is that most people are familiair with redirects.

The most "MySQL" like way is to use ```SELECT ... INTO OUTFILE``` but it isn't possible to write to a local file.

I think you'll have to see \> as a special command and not like some shell script extension.

meeuw commented:
Hmm I've noticed ```\.``` which also has it's roots in shell script.

tsroten commented:
I think the issue is less that `\.`, `\>`, or `\>>` are inspired by the shell, and more that `\>` and `\>` are used within (after) actual SQL queries. In other words, the worry is more about shell-like syntax, rather than commands named like shell equivalents.

The only special command that uses queries right now is `\fs`, which begins with the special command, so it's easily parse-able. There is no ambiguity.

I agree with @amjith that is seems harmless. I'm worried about parsing edge cases.

It's interesting that both psql and sqlite3 implement this feature in separate commands:
- `.output` or `.once` for [sqlite3](https://sqlite.org/cli.html#writing_results_to_a_file)
- `\o` or `\out` for [psql](https://www.postgresql.org/docs/9.6/static/app-psql.html)

sqlite's `.once` is a neat idea because you don't have to disable it.

Plus, you can already do `>`/`>>` in a real shell with mycli:
```bash
mycli my_database --csv -e 'select id, first_name, last_name from users limit 5' > results.csv
```

meeuw commented:
The special command ```\e``` also uses queries (I've inspired this patch on the logic used for ```\e```)

But I could live with a new ```once``` special command, I'm not sure if it'll work. I will make a new PR/branch for that.
